src/Microsoft.Identity.Client/Internal/ClientCredentialWrapper.cs

```diff
@@ -17,41 +17,51 @@ namespace Microsoft.Identity.Client.Internal
-        /// </summary>
-        /// <param name="certificate">contains information about the certificate previously shared with AAD at application
-        /// registration to prove the identity of the application (the client) requesting the tokens.</param>
-        public ClientCredentialWrapper(ClientAssertionCertificateWrapper certificate)
+        public static ClientCredentialWrapper CreateWithCertificate(ClientAssertionCertificateWrapper certificate)
+        {
+           return new ClientCredentialWrapper(certificate);
+        }
+        public static ClientCredentialWrapper CreateWithSecret(string secret)
+        {
+            return new ClientCredentialWrapper(secret, false);
+        }
+        public static ClientCredentialWrapper CreateWithSignedClientAssertion(string signedClientAssertion)
+        {
+            return new ClientCredentialWrapper(signedClientAssertion, true);
+        }
+
+        private ClientCredentialWrapper(ClientAssertionCertificateWrapper certificate)
         {
             ConfidentialClientApplication.GuardMobileFrameworks();
             Certificate = certificate;
         }
 
-        internal ClientAssertionCertificateWrapper Certificate { get; private set; }
-        internal string Assertion { get; set; }
-        internal long ValidTo { get; set; }
-        internal bool ContainsX5C { get; set; }
-        internal string Audience { get; set; }
-
-        /// <summary>
-        /// Constructor of client (application) credentials from a client secret, also known as the application password.
-        /// </summary>
-        /// <param name="secret">Secret string previously shared with AAD at application registration to prove the identity
-        /// of the application (the client) requesting the tokens.</param>
-        public ClientCredentialWrapper(string secret)
+        private ClientCredentialWrapper(string secretOrAssertion, bool isSignedAssertion)
         {
             ConfidentialClientApplication.GuardMobileFrameworks();
 
-            if (string.IsNullOrWhiteSpace(secret))
+            if (string.IsNullOrWhiteSpace(secretOrAssertion))
             {
-                throw new ArgumentNullException(nameof(secret));
+                throw new ArgumentNullException(nameof(secretOrAssertion));
             }
 
-            Secret = secret;
+            if (isSignedAssertion)
+            {
+                SignedAssertion = secretOrAssertion;
+            }
+            else
+            {
+                Secret = secretOrAssertion;
+            }
         }
 
+        internal ClientAssertionCertificateWrapper Certificate { get; private set; }
+        internal string Assertion { get; set; }
+        internal long ValidTo { get; set; }
+        internal bool ContainsX5C { get; set; }
+        internal string Audience { get; set; }
         internal string Secret { get; private set; }
+        internal string SignedAssertion { get; set; }
```

Comment:
This is  confusing ... Assertion and SignedAssertion... what is the difference?

Reply:
good point, the Assertion parameter was originally used for caching the assertion created after the claims were signed with the provided certificate. The signed assertion is the assertion that is already signed and passed in by the user.
Ill change assertion to cachedAssertion